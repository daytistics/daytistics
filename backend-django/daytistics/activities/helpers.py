from ..activities.models import ActivityType
from ..users.models import CustomUser


def build_activity_type_response(activity_type: ActivityType, user: CustomUser) -> dict:
    """
    Build a response schema for an ActivityType. This function takes an ActivityType object and a CustomUser object and returns a dictionary with the response schema.

    Args:
        activity_type (ActivityType): The ActivityType object to build the response schema for.
        user (CustomUser): The CustomUser object to check if the ActivityType is available for the user.

    Returns:
        dict: The response schema for the ActivityType.
    """

    return {
        "id": activity_type.pk,
        "name": activity_type.name,
        "category": activity_type.category.name,
        "available": activity_type in user.activities.all(),
        "active": activity_type.active,
    }
