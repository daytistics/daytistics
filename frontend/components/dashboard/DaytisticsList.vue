<template>
    <DashboardAddDaytisticDialog
        :open="isAddingDaytistic"
        @close="isAddingDaytistic = false"
    />

    <div class="bg-white shadow overflow-hidden sm:rounded-lg">
        <div class="px-4 py-5 sm:px-6 flex flex-row gap-3">
            <h3 class="text-lg leading-6 font-medium text-gray-900">
                Your Daytistics
            </h3>
            <button
                @click="isAddingDaytistic = true"
                class="text-secondary hover:text-secondary-dark flex items-center transition duration-150 ease-in-out"
            >
                <Plus class="w-5 h-5 mr-1" />
                <span class="text-sm font-medium">Add New</span>
            </button>
        </div>
        <div class="border-t border-gray-200">
            <div class="px-4 py-5 sm:p-6">
                <div class="flex items-center justify-between mb-4">
                    <input
                        type="text"
                        placeholder="Search date..."
                        class="shadow-sm focus:ring-indigo-500 focus:border-indigo-500 block w-full sm:text-sm border-gray-300 rounded-md"
                    />
                    <button
                        class="ml-3 inline-flex items-center px-4 py-2 border border-transparent text-sm font-medium rounded-md text-white bg-secondary hover:bg-secondary-dark"
                    >
                        Search
                    </button>
                </div>
                <ul class="divide-y divide-gray-200">
                    <li
                        v-for="(daytistic, index) in daytistics"
                        :key="index"
                        class="py-4 cursor-pointer hover:bg-gray-50"
                    >
                        <NuxtLink :to="`/daytistics/${daytistic.id}`">
                            <div class="flex items-center space-x-4">
                                <Calendar class="h-6 w-6 text-gray-400" />
                                <div class="flex-1 min-w-0">
                                    <p
                                        class="text-sm font-medium text-gray-900 truncate"
                                    >
                                        {{
                                            new Date(
                                                daytistic.date
                                            ).toLocaleDateString()
                                        }}
                                    </p>
                                    <p class="text-sm text-gray-500 truncate">
                                        {{ daytistic.total_activities }}
                                        Activities /
                                        {{ daytistic.total_duration }}
                                    </p>
                                </div>
                            </div>
                        </NuxtLink>
                    </li>
                    <div class="flex justify-center">
                        <span
                            v-show="totalPages === 0"
                            class="text-lg"
                        >
                            We couldn't find any daytistics for you. Why not
                            <button
                                type="button"
                                class="text-secondary"
                            >
                                create
                            </button>
                            one?
                        </span>
                    </div>
                </ul>
                <div class="mt-4 flex items-center justify-between">
                    <button
                        :disabled="currentPage === 1"
                        @click="currentPage--"
                        class="relative inline-flex items-center px-4 py-2 border border-gray-300 text-sm font-medium rounded-md text-gray-700 bg-white hover:bg-gray-50"
                    >
                        Previous
                    </button>
                    <span
                        v-show="totalPages > 0"
                        class="text-sm text-gray-700"
                    >
                        Page {{ currentPage }} of {{ totalPages }}
                    </span>
                    <button
                        :disabled="currentPage === totalPages"
                        @click="currentPage++"
                        class="relative inline-flex items-center px-4 py-2 border border-gray-300 text-sm font-medium rounded-md text-gray-700 bg-white hover:bg-gray-50"
                    >
                        Next
                    </button>
                </div>
            </div>
        </div>
    </div>
</template>

<script lang="ts" setup>
import { Calendar, Plus } from 'lucide-vue-next';
import type { Daytistic } from '~/types/daytistics';
const daytistics = ref<Daytistic[] | []>([]);

const listDaytisticsAPI = useListDaytisticsAPI();

const currentPage = ref(1);
const totalPages = ref(1);
const isAddingDaytistic = ref(false);

watch(currentPage, async (newPage) => {
    if (totalPages.value > 0) {
        await listDaytisticsAPI.loadPage(newPage);
    }
});

onMounted(async () => {
    await listDaytisticsAPI.loadPage(currentPage.value);
});

function useListDaytisticsAPI() {
    const { $api } = useNuxtApp();

    const list = async (page: number) => {
        return $api(`/api/daytistics/list?page=${page}`, {
            server: false,
            method: 'GET',
        });
    };

    const loadPage = async (page: number) => {
        const response: any = await list(page);
        daytistics.value = response.items;
        totalPages.value = Math.ceil(response.count / 5);
    };
    return {
        loadPage,
    };
}
</script>
