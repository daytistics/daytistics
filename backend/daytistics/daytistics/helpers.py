from typing import Dict, Union, List

from .models import Daytistic
from ..users.models import CustomUser
from ..activities.models import ActivityEntry


def build_daytistic_response(
    daytistic: Daytistic,
) -> dict:
    """
    Build a response schema for a Daytistic. This function takes a Daytistic object and returns a dictionary with the response schema.

    Args:
        daytistic (Daytistic): The Daytistic object to build the response schema for.
    Returns:
        dict: The response schema for the Daytistic.
    """

    user = CustomUser.objects.get(id=daytistic.user.pk)

    last_login = (
        user.last_login.strftime("%Y-%m-%dT%H:%M:%S") if user.last_login else "Never"
    )

    user_schema = {
        "username": user.username,
        "email": user.email,
        "is_active": user.is_active,
        "is_staff": user.is_staff,
        "is_superuser": user.is_superuser,
        "groups": [group.name for group in user.groups.all()],
        "user_permissions": [
            permission.codename for permission in user.user_permissions.all()
        ],
        "date_joined": user.date_joined.strftime("%Y-%m-%dT%H:%M:%S"),
        "last_login": last_login,
        "date_format": user.date_format,
    }

    diary_schema = {
        "entry": "Lorem ipsum dolor sit amet, consectetur adipiscing elit. Nullam nec nunc nec nunc.",
        "moment_of_happiness": "Lorem ipsum dolor sit amet, consectetur adipiscing elit. Nullam nec nunc nec nunc.",
    }

    wellbeing_schema_list = [
        {
            "id": i,
            "name": f"Well-being #{i}",
            "rating": i - 1,
        }
        for i in range(1, 7)
    ]

    activity_schema_list = [
        build_activity_response(activity)
        for activity in daytistic.activities.all().order_by("start_time")
    ]

    # Converts the date to a datetime object
    daytistic_datetime = daytistic.date.strftime("%Y-%m-%d")

    return {
        "id": daytistic.pk,
        "date": daytistic_datetime,
        "average_wellbeing": 5.0,
        "total_activities": daytistic.activities.count(),
        "total_duration": daytistic.total_duration,
        "user": user_schema,
        "wellbeing": wellbeing_schema_list,
        "activities": activity_schema_list,
        "diary": diary_schema,
    }


def build_activity_response(activity: ActivityEntry) -> Dict[str, Union[int, str]]:
    """
    Build a response schema for an ActivityEntry. This function takes an ActivityEntry object and returns a dictionary with the response schema.

    Args:
        activity (ActivityEntry): The ActivityEntry object to build the response schema for.

    Returns:
        Dict[str, Union[int, str]]: The response schema for the ActivityEntry.
    """

    return {
        "id": activity.pk,
        "name": activity.type.name,
        "duration": activity.duration,
        "start_time": activity.start_time,
        "end_time": activity.end_time,
    }
