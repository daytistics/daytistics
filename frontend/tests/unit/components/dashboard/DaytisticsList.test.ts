import { mount } from '@vue/test-utils';
import { describe, it, expect, beforeEach, vitest, vi } from 'vitest';
import DaytisticsList from '~/components/dashboard/DaytisticsList.vue';
import { mountSuspended } from '@nuxt/test-utils/runtime';
import { resolve } from 'chart.js/helpers';

describe('YourDaytistics Component', () => {
    let wrapper: any;

    beforeEach(async () => {
        wrapper = await mountSuspended(DaytisticsList, {
            props: {
                daytistics: [
                    {
                        id: 1,
                        date: '2024-10-19',
                        total_activities: 3,
                        total_duration: 120,
                    },
                ],
                totalPages: 1,
            },
        });
    });

    it('renders daytistics correctly', () => {
        const date = new Date('2024-10-19');
        expect(wrapper.text()).toContain('Your Daytistics');
        expect(wrapper.text()).toContain(date.toLocaleDateString());
        expect(wrapper.text()).toContain('3 Activities / 2h');
    });

    it('shows a message when no daytistics are found', async () => {
        await wrapper.setProps({ daytistics: [] });
        expect(wrapper.text()).toContain(
            "We couldn't find any daytistics for you."
        );
    });

    // it('calls the API when changing the page', async () => {
    //     const mockLoadPage = vitest.spyOn(wrapper.vm, 'loadPage');
    //     await wrapper.vm.currentPage = 2;
    //     wrapper.vm.$nextTick();
    //     expect(mockLoadPage).toHaveBeenCalledWith(2);
    // });
});
