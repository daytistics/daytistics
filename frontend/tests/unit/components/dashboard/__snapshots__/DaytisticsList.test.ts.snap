// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`DaytisticsList > should render the component 1`] = `
"<div class="bg-white shadow overflow-hidden sm:rounded-lg">
  <div class="px-4 py-5 sm:px-6 flex flex-row gap-3">
    <h3 class="text-lg leading-6 font-medium text-gray-900"> Your Daytistics </h3><button data-modal-target="add-daytistic-modal" data-modal-show="add-daytistic-modal" class="text-secondary hover:text-secondary-dark flex items-center transition duration-150 ease-in-out"><svg xmlns="http://www.w3.org/2000/svg" width="24" height="24" viewBox="0 0 24 24" fill="none" stroke="currentColor" stroke-width="2" stroke-linecap="round" stroke-linejoin="round" class="lucide lucide-plus-icon w-5 h-5 mr-1">
        <path d="M5 12h14"></path>
        <path d="M12 5v14"></path>
      </svg><span class="text-sm font-medium">Add New</span></button>
  </div>
  <div class="border-t border-gray-200">
    <div class="px-4 py-5 sm:p-6">
      <div class="flex items-center justify-between mb-4"><input type="text" placeholder="Search date..." class="shadow-sm focus:ring-indigo-500 focus:border-indigo-500 block w-full sm:text-sm border-gray-300 rounded-md"><button class="ml-3 inline-flex items-center px-4 py-2 border border-transparent text-sm font-medium rounded-md text-white bg-secondary hover:bg-secondary-dark"> Search </button></div>
      <ul class="divide-y divide-gray-200">
        <div class="flex justify-center"><span class="text-lg" style="display: none;"> We couldn't find any daytistics for you. Why not <button data-modal-target="add-daytistic-modal" data-modal-show="add-daytistic-modal" type="button" class="text-secondary"> create </button> one? </span></div>
      </ul>
      <div class="mt-4 flex items-center justify-between"><button disabled="" class="relative inline-flex items-center px-4 py-2 border border-gray-300 text-sm font-medium rounded-md text-gray-700 bg-white hover:bg-gray-50"> Previous </button><span class="text-sm text-gray-700"> Page 1 of 1</span><button disabled="" class="relative inline-flex items-center px-4 py-2 border border-gray-300 text-sm font-medium rounded-md text-gray-700 bg-white hover:bg-gray-50"> Next </button></div>
    </div>
  </div>
</div>"
`;
